tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp

```diff
@@ -0,0 +1,598 @@
+#include <olp/core/utils/Dir.h>
+#include <olp/dataservice/read/PrefetchTilesRequest.h>
+#include <testutils/CustomParameters.hpp>
+
+#include "olp/dataservice/read/CatalogClient.h"
+#include "olp/dataservice/read/CatalogRequest.h"
+#include "olp/dataservice/read/CatalogVersionRequest.h"
+#include "olp/dataservice/read/DataRequest.h"
+#include "olp/dataservice/read/PartitionsRequest.h"
+#include "olp/dataservice/read/model/Catalog.h"
+
+using namespace olp::dataservice::read;
+using namespace testing;
+
+namespace {
+
+enum class CacheType { IN_MEMORY, DISK, BOTH };
+
+std::ostream& operator<<(std::ostream& os, const CacheType cache_type) {
+  switch (cache_type) {
+    case CacheType::IN_MEMORY:
+      return os << "In-memory cache";
+    case CacheType::DISK:
+      return os << "Disk cache";
+    case CacheType::BOTH:
+      return os << "In-memory & disk cache";
+    default:
+      return os << "Unknown cache type";
+  }
+}
+
+}  // namespace
+
+class CatalogClientTest : public ::testing::TestWithParam<CacheType> {
+ public:
+  void SetUp() override {
+    auto network = olp::client::OlpClientSettingsFactory::
+        CreateDefaultNetworkRequestHandler();
+
+    olp::authentication::Settings auth_settings;
+    auth_settings.network_request_handler = network;
+
+    olp::authentication::TokenProviderDefault provider(
+        CustomParameters::getArgument("dataservice_read_test_appid"),
+        CustomParameters::getArgument("dataservice_read_test_secret"),
+        auth_settings);
+    olp::client::AuthenticationSettings auth_client_settings;
+    auth_client_settings.provider = provider;
+    settings_ = std::make_shared<olp::client::OlpClientSettings>();
+    settings_->network_request_handler = network;
+    settings_->authentication_settings = auth_client_settings;
+    client_ = olp::client::OlpClientFactory::Create(*settings_);
+  }
+
+  void TearDown() override {
+    client_.reset();
+    auto network = std::move(settings_->network_request_handler);
+    settings_.reset();
+    // when test ends we must be sure that network pointer is not captured
+    // anywhere
+    assert(network.use_count() == 1);
+  }
+
+ protected:
+  std::string GetTestCatalog() {
+    return CustomParameters::getArgument("dataservice_read_test_catalog");
+  }
+
+  std::string PrintError(const olp::client::ApiError& error) {
+    std::ostringstream resultStream;
+    resultStream << "ERROR: code: " << static_cast<int>(error.GetErrorCode())
+                 << ", status: " << error.GetHttpStatusCode()
+                 << ", message: " << error.GetMessage();
+    return resultStream.str();
+  }
+
+  template <typename T>
+  T GetExecutionTime(std::function<T()> func) {
+    auto startTime = std::chrono::high_resolution_clock::now();
+    auto result = func();
+    auto end = std::chrono::high_resolution_clock::now();
+    std::chrono::duration<double> time = end - startTime;
+    std::cout << "duration: " << time.count() * 1000000 << " us" << std::endl;
```

Comment:
Do we need to log the duration of function execution? I suggest to remove it in order to not clutter the log.

Reply:
I'd prefer to not change the test's behaviour in this PR. This part was implemented intentionally, at least requires more discussion before cutting it off.

Reply:
What about usage of `OLP_SDK_LOG_TRACE` in this case? The `std::cout` won't produce anything on the Android platform.